File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.cpp
/*
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

Node* newNode(int val) {
    return new Node(val);
}
*/

class Solution {
  public:
    int dfs(Node* root, int &ans) {
        if(!root) return 0;
        int left = max(0,dfs(root->left,ans));
        int right = max(0,dfs(root->right,ans));
        ans = max(ans,right+left);
        
        return 1+max(left,right);
        
    }
    int diameter(Node* root) {
        // Your code here
        int ans = 0;
         dfs(root,ans);
         return ans;
    }
};

File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.java
/*
Definition for Node
class Node {
    int data;
    Node left;
    Node right;
    Node(int data) {
        this.data = data;
        left = right = null;
    }
} */

class Solution {
    public int solve(Node root,int[] ans) {
        if(root==null) return 0;
        
        int left = solve(root.left,ans);
        
        int right = solve(root.right,ans);
        
        ans[0] = Math.max(ans[0],left+right);
        
        return Math.max(left,right)+1;
    }
    public int diameter(Node root) {
        // code here
        int[] ans = new int[1];
        ans[0] = 0;
        int x = solve(root,ans);
        //System.out.println(x);
        return ans[0];
    }
}